File: p159h/longest_substr_two_distinct.py
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: str) -> int:
        char_counter = 256 * [0]
        ans, start, distinct = 0, 0, 0
        for end in range(len(s)):
            if char_counter[ord(s[end])] == 0:
                distinct += 1
            char_counter[ord(s[end])] += 1
            while distinct > 2:
                char_counter[ord(s[start])] -= 1
                if char_counter[ord(s[start])] == 0:
                    distinct -= 1
                start += 1
            ans = max(ans, end - start + 1)
        return ans


## TESTS
tests = [
    ("eceba", 3),
    ("ccaabbb", 5),
    ("CCCCC", 5),
    ("abcde", 2),
    ("CODEaaaabbbbcde", 8),
]
for t in tests:
    sol = Solution()
    actual = sol.lengthOfLongestSubstringTwoDistinct(t[0])
    print(
        "Longest substring length in '"
        + t[0]
        + "' with at most two distincts ->",
        actual,
    )
    assert actual == t[1]

File: p159h/longest_substr_two_distinct.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstringTwoDistinct = function(s) {
  const charCounter = Array.from({ length: 128 }, v => 0);
  let [ans, start, distinct] = [0, 0, 0];
  for (let end = 0; end < s.length; ++end) {
    // move end pointer and update hashmap's char occurences
    if (charCounter[s.charCodeAt(end)]++ === 0) distinct++;
    // until the distinct exceeding 2, then we move start to drop the distinct
    while (distinct > 2) {
      if (charCounter[s.charCodeAt(start)]-- === 1) distinct--;
      start++;
    }
    ans = Math.max(ans, end - start + 1);
  }
  return ans;
};
// TEST
[
  ['eceba', 3],
  ['ccaabbb', 5],
  ['CCCCC', 5],
  ['abcde', 2],
  ['CODEaaaabbbbcde', 8]
].forEach(t => {
  actual = lengthOfLongestSubstringTwoDistinct(t[0]);
  console.log(
    "Longest substring length in '" + t[0] + "' with at most two distincts ->",
    actual
  );
  console.assert(t[1] === actual);
});
